### ISSUE TITLE
Readiness probe failed  8181: connect: connection refused

### ISSUE THREAD

**Issue by xieyanker** [coredns/coredns#3219]
Hello, my coredns's version is 1.5.0, I deploy it in the k8s.
And I deploy coredns by deployment, its' yaml has livenessProbe with ReadinessProbe.
But the pod's ReadinessProbe is often failed.

```shell
[root@master3 log]# kubectl get pod -nkube-system -owide|grep coredns
coredns-58f78cf5f4-268g6                   0/1     Running   0          2d4h    10.151.164.145   master1   <none>           <none>
coredns-58f78cf5f4-2cfwx                   0/1     Running   0          24h     10.151.208.191   master2   <none>           <none>
coredns-58f78cf5f4-nndv8                   0/1     Running   0          6h32m   10.151.32.112    master3   <none>           <none>
```

I describe the pod's event as follows:
```shell
Events:
  Type     Reason            Age                     From              Message
  ----     ------            ----                    ----              -------
  Warning  DNSConfigForming  5m25s (x305 over 6h2m)  kubelet, master3  Nameserver limits were exceeded, some nameservers have been omitted, the applied nameserver line is: 172.16.3.1 172.16.3.2 172.16.3.5
  Warning  Unhealthy         59s (x85 over 13m)      kubelet, master3  Readiness probe failed: Get http://10.151.32.112:8181/ready: dial tcp 10.151.32.112:8181: connect: connection refused
```

I try to connect the pod's 8181 port by using telnet, but failed.
```shell
[root@master1 ~]# telnet  10.151.164.145 8181
Trying 10.151.164.145...
telnet: connect to address 10.151.164.145: Connection refused
[root@master1 ~]# telnet  10.151.164.145 8080
Trying 10.151.164.145...
Connected to 10.151.164.145.
Escape character is '^]'.
^]
```

my readinessProbe and livenessProbe as follows:
```yaml
    readinessProbe:
      failureThreshold: 10
      httpGet:
        path: /ready
        port: 8181
        scheme: HTTP
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 5

    livenessProbe:
      failureThreshold: 10
      httpGet:
        path: /health
        port: 8080
        scheme: HTTP
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 5
```

**Comment by chrisohaver** [comment]
What do the coredns pod logs show?

And please share your Corefile.

**Comment by xieyanker** [comment]
@chrisohaver 
Hi, This is my Corefile:
```yaml
    .:53 {
        errors
        health
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          upstream /etc/resolv.conf
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . 172.16.3.1 {
          policy sequential
          prefer_udp
        }
        cache 30
        loop
        reload
        loadbalance
    }
```
I have restarted the pod by force, I cannot get the log at the time.
And I found the `8181` port is only check the plugin which have implemented the `Readiness interface`. However, all of the plugins which I used not implement it. 
So I am planning to remove the readinessProbe, and keep only the livenessProbe.

**Comment by chrisohaver** [comment]
>I have restarted the pod by force, I cannot get the log at the time.

You can usually get it with the `-p` option, which gets the log for _previous_ instance.

> And I found the 8181 port is only check the plugin which have implemented the Readiness interface. However, all of the plugins which I used not implement it.

The _kubernetes_ plugin implements it, which you are using.

**Comment by chrisohaver** [comment]
Try upgrading to a more recent version of CoreDNS (like 1.6.2) - I think there is a ready plugin bug in v1.5.0 that causes it to not listen ...

#2814 - which looks like a fix related to config reloads - but may contain other fixes (like default listening port?).

**Comment by xieyanker** [comment]
@chrisohaver Thanks for your reply, but I have removed the previous pod form container. I will restore the readinessProbe, and I will keep an eye on it.

**Comment by cann0nf0dder** [comment]
not sure if this helps but I'm running 1.5.0 with similar config and it works fine. I'd recommend to look at https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed to see if there are any discrepancies. My liveness config matches the one in the url and i didn't see it in yours (as example):
`  initialDelaySeconds: 60`

**Comment by xieyanker** [comment]
@cann0nf0dder I think my config is normal, my liveness is ok, the error is readiness.

**Comment by chrisohaver** [comment]
Knowing what is in the coredns pod logs could help.

**Comment by chrisohaver** [comment]
reopen if it's still an issue.

**Comment by xieyanker** [comment]
@chrisohaver 
Thank you, I have found the origin of this issue, the `coredns` pod will check readiness failed when I edit `coredns`'s `configmap` add a plugin to `Corefile`. If I don't delete the `coredns` pod, it will reload config and check readiness failed after serval minutes.

> Try upgrading to a more recent version of CoreDNS (like 1.6.2) - I think there is a ready plugin bug in v1.5.0 that causes it to not listen ...
>
> #2814 - which looks like a fix related to config reloads - but may contain other fixes (like default listening port?).

This can resolve my issue, it will not happen when I use the coredns 1.6.4 version.

**Comment by yonahd** [comment]
@xieyanker 
I have solved this by adding ready in both parts of the corefile.
“Lessons learned from upgrading bare metal kubernetes offline to 1.16” 
https://link.medium.com/j6FVuVBxv1

**Comment by xieyanker** [comment]
@yonahd your issue is different to mine, I have added the `ready` plugin. The reason of mine is my `coredns`'s version is too low, it will get `0/1` if I don't restart the pod after I update the configmap.

**Comment by yonahd** [comment]
@xieyanker The default kubernetes wants is 1.6.2 although now I see you have 1.5.0 which had failed reloads fixed in 1.5.1 
https://coredns.io/2019/06/26/coredns-1.5.1-release/

**Comment by ykfq** [comment]
Update coredns image from v1.5.0 to current version v1.6.9, then the error got fixed.
